src/main/java/org/apache/sysds/runtime/util/DataConverter.java

```diff
@@ -307,6 +307,34 @@ public static MatrixBlock readMatrixFromHDFS(ReadProperties prop)
 		return ret;
 	}
 
+	public static List<Integer> convertVectorToIndicesList(MatrixBlock mb)
+	{
+		int rows = mb.getNumRows();
+		int cols = mb.getNumColumns();
+		ArrayList<Integer> indices = new ArrayList<>();
+
+		if( mb.getNonZeros() > 0 )
+		{
+			if( mb.isInSparseFormat() )
+			{
+				Iterator<IJV> iter = mb.getSparseBlockIterator();
+				while( iter.hasNext() ) {
+					IJV cell = iter.next();
+					if( cell.getV() != 0.0)
+						indices.add(cell.getI()*cols+cell.getJ());
+				}
+			}
+			else {
+				for( int i=0, cix=0; i<rows; i++ )
+					for( int j=0; j<cols; j++, cix++)
+						if( mb.getValueDenseUnsafe(i, j) != 0.0 )
+							indices.add(cix);
```

Comment:
don't add to the ArrayList (in general don't use an ArrayList here)

preallocate a single int[] based on nnz in the input.
then put the values into that.